#167377525 Setup sequelize database

successgilli opened the pull request:
#### What does this PR do?
This pull request initialises the database
#### Description of Task to be completed?
- install sequelize and sequelize-cli
- provide basic setup for the database for development, production and test environments
#### How should this be manually tested?
- git clone this branch
- run 'npm run db:roll' which will create a users table in your local database
#### Any background context you want to provide?
Ensure to have the right information in your .env file as is in the .env.example file
#### What are the relevant pivotal tracker stories?
[#167377525](https://www.pivotaltracker.com/story/show/167377525)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

Commit:
chore(database): setup database

- installed sequelize
- created database structure

[starts #167377525]

Commit:
chore(database): setup database

- installed sequelize
- created database structure

[starts #167377525]

Commit:
chore(database): setup database

- installed sequelize, sequelize-cli and @babel/register
- created database structure
- added babelHook.js to allow config to use es6
- removed mongodb's models folder from root
[Starts #167377525]

Commit:
chore(database): setup database

- installed sequelize, sequelize-cli and @babel/register
- created database structure
- added babelHook.js to allow config to use es6
- removed mongodb's models folder from root
[Starts #167377525]

iverenshaguy commented:
Please fix all these hound errors and ensure that your commit message format and pull request description adhere to my reviews in other PRs

successgilli commented:
> Please fix all these hound errors and ensure that your commit message format and pull request description adhere to my reviews in other PRs

@iverenshaguy, Thanks for the comment. I actually used the Es5 syntax because babel was not already setup yet and most of the parts of my code were generated by running sequelize migrations. Also, the sequelize codes that were generated by migrations has refused to run properly when ever I try to change the Es5 to Es6 even when I took the app for a spin with babel.
I will have to wait for my colleague to merge his babel setup then I will do the conversion to Es6.

iverenshaguy commented:
@successgilli Any progress on this? I informed you that I merged in the Babel chore around 3:30pm yesterday but I can't see any extra work done after that. What's going on?
<img width="399" alt="Screenshot 2019-07-24 at 07 29 36" src="https://user-images.githubusercontent.com/22734663/61770295-d4738500-ade4-11e9-8d4c-8a36bce9309d.png">

kennyedward reviewed (approved):
All checks passed. This PR looks ready for merging.

iverenshaguy commented:
Please update your PR description to `This pull request` not `This PR` and rebase on develop.

successgilli commented:
> Please update your PR description to `This pull request` not `This PR` and rebase on develop.

okay @iverenshaguy

Jonathan4github reviewed (approved):
Good job.

ibkadeeko reviewed (approved):
Looks good to me

iverenshaguy commented on package.json:
```diff
@@ -4,18 +4,21 @@
   "scripts": {
+    "build": "babel src --out-dir dist",
+    "coveralls": "cat ./coverage/lcov.info | coveralls",
+    "db:roll": "sequelize db:migrate",
```
`db:roll`??? Please change to `db:migrate`. Also, instead of using babelHook which will need us to move @babel/register to dependencies instead of devDependencies, I believe we should use my solution i.e
```suggestion
    "sequelize-cli:es6": "babel-node ./node_modules/.bin/sequelize $*"
```
This uses babel-node to run sequelize therefore transpiling it on each run.

Then the following scripts will be upgraded to

```
    "db:migrate": "npm run sequelize-cli:es6 db:migrate",
    "db:undo": "npm run sequelize-cli:es6 db:migrate:undo:all",
    "db:setup": "npm run db:migrate && npm run sequelize-cli:es6 db:seed:all",
```

In the config, we'll now have

```
'config': path.resolve('./config/sequelize'),
```

successgilli replied:
I see your point now

iverenshaguy commented on package.json:
```diff
@@ -4,18 +4,21 @@
   "author": "Andela Simulations Programme",
   "license": "MIT",
   "dependencies": {
+    "@babel/register": "^7.4.4",
```
Please remove this since we will now use `babel-node`.

iverenshaguy commented on package.json:
```diff
@@ -27,12 +30,14 @@
+    "pg-hstore": "^2.3.3",
     "request": "^2.87.0",
+    "sequelize": "^5.10.2",
+    "sequelize-cli": "^5.5.0",
```
This should be a dev dependency. Only sequelize itself is a main dependency. Please move it to it's correct position

successgilli replied:
okay, I will do this.

iverenshaguy commented on src/db/config/config.js:
```diff
@@ -0,0 +1,28 @@
+import { config } from 'dotenv';
```
I believe we should use `DATABASE_URL` environmental variable for our config. This format of `username: process.env.DB_USERNAME_DEV etc`  is just too messy. We should also rename this file to `sequelize.js`, `config/config.js` seems really weird.

So we can replace it with the following

```
import { config } from 'dotenv';

config();

const databaseUrls = {
  development: process.env.DATABASE_URL,
  staging: process.env.DATABASE_URL,
  test: process.env.DATABASE_TEST_URL,
  production: process.env.DATABASE_URL
};

const environment = process.env.NODE_ENV || 'development';
const url = databaseUrls[environment];
const devMode = (environment !== 'production');

const config = {
  url,
  dialect: 'postgres,
  logging: devMode ? log => log : false,
  operatorsAliases: false,
};

export default config;
```

This is more concise and easy to think through

iverenshaguy replied:
The above also logs the database queries whenever we are in dev mode which makes it easier to debug instead of totally turning off the logger.

iverenshaguy commented on src/db/migrations/20190723163335-create-user.js:
```diff
@@ -0,0 +1,29 @@
+
+export default {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Users', {
```
Please use async-await and try/catch. i.e 
```suggestion
  async up(queryInterface, Sequelize) { 
    try {
      queryInterface.createTable('Users', {...
    } catch (error) {
      console.log(error);
    }
```

This will help us to easily catch errors that occur during migration. Please do the same thing for `down`.

iverenshaguy commented on src/db/migrations/20190723163335-create-user.js:
```diff
@@ -0,0 +1,29 @@
+
+export default {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('Users', {
+    id: {
```
We would be using `UUID(v4)` as our primary keys.

iverenshaguy commented on src/db/models/index.js:
```diff
@@ -0,0 +1,35 @@
+import Sequelize from 'sequelize';
+import fs from 'fs';
+import path from 'path';
+import * as dbconfig from '../config/config';
```
```suggestion
import config from '../config/sequelize';
```

iverenshaguy commented on src/db/models/index.js:
```diff
@@ -0,0 +1,35 @@
+import * as dbconfig from '../config/config';
+
+const basename = path.basename(__filename);
+const env = process.env.NODE_ENV || 'development';
```
```suggestion
```

iverenshaguy commented on src/db/models/index.js:
```diff
@@ -0,0 +1,35 @@
+
+const basename = path.basename(__filename);
+const env = process.env.NODE_ENV || 'development';
+const config = dbconfig[env];
```
```suggestion
```

iverenshaguy commented on src/db/models/index.js:
```diff
@@ -0,0 +1,35 @@
+const config = dbconfig[env];
+const db = {};
+
+let sequelize;
```
```suggestion
const sequelize = new Sequelize(config.url, config);
```

iverenshaguy replied:
Remove the if/else logic

iverenshaguy commented on src/db/models/index.js:
```diff
@@ -0,0 +1,35 @@
+    db[model.name] = model;
+  });
+
+Object.keys(db).forEach((modelName) => {
```
model not modelName
```suggestion
Object.keys(db).forEach((model) => {
```

iverenshaguy commented on src/db/models/index.js:
```diff
@@ -0,0 +1,35 @@
+db.sequelize = sequelize;
+db.Sequelize = Sequelize;
+
+export default db;
```
Please change all `db` to `database`.

iverenshaguy commented on src/db/models/user.js:
```diff
@@ -0,0 +1,14 @@
+
```
Rename the file to `user.model.js`

iverenshaguy commented on src/db/models/user.js:
```diff
@@ -0,0 +1,14 @@
+    lastName: DataTypes.STRING,
+    email: DataTypes.STRING
+  }, {});
+  User.associate = () => {
```
Please leave a new line above lines 8 and 11

iverenshaguy commented on src/index.js:
```diff
@@ -35,30 +35,21 @@ if (!isProduction) {
 app.use(require('./routes'));
 
-// / catch 404 and forward to error handler
+// catch 404 and forward to error handler
```
Remove the weird comments on lines 40, 47 and 76

iverenshaguy commented on .sequelizerc:
```diff
@@ -0,0 +1,9 @@
+const path = require('path');
+
+module.exports = {
+  'config': path.resolve('./src/db/config', 'babelHook.js'),
```
Remove config from the src folder.
```suggestion
  'config': path.resolve('./config/sequelize', 'babelHook.js'),
```

iverenshaguy commented on .sequelizerc:
```diff
@@ -0,0 +1,9 @@
+
+module.exports = {
+  'config': path.resolve('./src/db/config', 'babelHook.js'),
+  'models-path': path.resolve('src/db', 'models'),
```
Models should be in the src folder, not DB folder. We are not using a DB folder.
```suggestion
  'models-path': path.resolve('./src', 'models'),
```

iverenshaguy commented on .sequelizerc:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  'config': path.resolve('./src/db/config', 'babelHook.js'),
+  'models-path': path.resolve('src/db', 'models'),
+  'seeders-path': path.resolve('src/db', 'seeders'),
```
Seeders and Migrations can be in the config folder outside the src folder. This is to avoid us transpiling code we do not need.
```suggestion
  'seeders-path': path.resolve('./config', 'seeders'),
```

iverenshaguy commented on .sequelizerc:
```diff
@@ -0,0 +1,9 @@
+  'config': path.resolve('./src/db/config', 'babelHook.js'),
+  'models-path': path.resolve('src/db', 'models'),
+  'seeders-path': path.resolve('src/db', 'seeders'),
+  'migrations-path': path.resolve('src/db', 'migrations')
```
```suggestion
  'migrations-path': path.resolve('./config, 'migrations')
```

iverenshaguy commented on package.json:
```diff
@@ -4,18 +4,21 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
```
Good job arranging this alphabetically

iverenshaguy commented on package.json:
```diff
@@ -4,18 +4,21 @@
+    "build": "babel src --out-dir dist",
+    "coveralls": "cat ./coverage/lcov.info | coveralls",
+    "db:roll": "sequelize db:migrate",
+    "db:undo": "sequelize db:migrate:undo:all",
```
We can add a `db:setup` script
```suggestion
    "db:undo": "sequelize db:migrate:undo:all",
    "db:setup": "npm run db:migrate && sequelize db:seed:all",
```

iverenshaguy commented on src/db/config/babelHook.js:
```diff
@@ -0,0 +1,3 @@
+require('@babel/register');
```
We don't need this file anymore. I realised that it would mean us adding `@babel-register` to our main dependencies which is not very efficient since it is not needed for the main code execution only for setup.

successgilli replied:
Okay.
I will make the necessary file changes and push.

iverenshaguy reviewed (changes requested):
Good job. Please review my comments and update

iverenshaguy committed:
chore(database): update database setup

- update sequelize config
- update config folder structure

[Starts #167377525]

Commit:
chore(database): update database setup

- added async to migrations files
- added try/catch block to migrations file
- refactored database setup

[Starts #167377525]

iverenshaguy committed:
chore(database): update database setup

- update sequelize config
- update config folder structure

[Starts #167377525]

Commit:
chore(database): update database setup

- added async to migration file
- added try/catch block in migrations

[Starts #167377525]

iverenshaguy committed:
chore(database): update database setup

- update .env.example file

[Starts #167377525]
